### ISSUE TITLE
[Bug] Handle error when printer is turned off 

### ISSUE THREAD

**Issue by greghesp** [greghesp/ha-bambulab#10]
```
This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/bambu_lab/pybambu/bambu_client.py:52
Integration: Bambu Lab (documentation, issues)
First occurred: 10 February 2023 at 22:58:40 (2819 occurrences)
Last logged: 06:48:20

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/bambu_lab/coordinator.py", line 46, in listen
    self.client.connect(callback=message_handler)
  File "/config/custom_components/bambu_lab/pybambu/bambu_client.py", line 52, in connect
    self.client.connect(self.host, self._port)
  File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 914, in connect
    return self.reconnect()
  File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1044, in reconnect
    sock = self._create_socket_connection()
  File "/usr/local/lib/python3.10/site-packages/paho/mqtt/client.py", line 3685, in _create_socket_connection
    return socket.create_connection(addr, timeout=self._connect_timeout, source_address=source)
  File "/usr/local/lib/python3.10/socket.py", line 845, in create_connection
    raise err
  File "/usr/local/lib/python3.10/socket.py", line 833, in create_connection
    sock.connect(sa)
OSError: [Errno 113] Host is unreachable
```

**Comment by AdrianGarside** [comment]
Should be easy to fix by wrapping the connection attempt in coordinator.py in a try/except - it'll keep attempting to connect every 10 seconds per the current polling rate. And once the printer is back on it reconnects successfully.

Reproing this now it took a long time before turning my P1P off triggered the MQTT lib to trigger disconnection. I didn't time it but maybe there's a keep alive we can reduce so it's detected earlier.

**Comment by AdrianGarside** [comment]
Be nice to add a sensor for the printer being unreachable/offline?

**Comment by AdrianGarside** [comment]
My changes to move the processsing to a background thread and use loop_forever means it automatically attempts to reconnect if the connection drops. So the work to detect and reconnect in the 10s poll is unnecessary (and harmful because now I spin up a new thread for each connection attempt).

**Comment by AdrianGarside** [comment]
Moving the initial connection into the coordinator constructor (this may breaks the initial config flow), setting the keepalive to 5s and relying on the automatic reconnect in the mqtt broker fixes this exception and gives a much more responsive reconnect when the printer is powered back on.

**Comment by AdrianGarside** [comment]
Fixed with https://github.com/greghesp/ha-bambulab/pull/15

**Comment by AdrianGarside** [comment]
I understand the concepts but still learning how HA likes things to be done. I don't think I need to change anything but what I've got now in #15 is working really well. I just finished merging that change with your lights topic branch so I could add a push_all command. So now when HA restarts, I get straight to fully populated and with real time updating of the sensors:
![image](https://user-images.githubusercontent.com/26147576/218400255-c04beb9d-a639-4dd9-913d-2de2333a7367.png)

**Comment by AdrianGarside** [comment]
Yes, we need to do an async initialization of the device  so we can wait on the version data to be returned before we tell HA the device is available. But the HA docs are little lacking in describing exactly how you do that. The async init stuff I've been reading up on is about async initializing a 'platform' and the async methods it describes don't match the ones we're currently using and I haven't been able to find much documentation for those.

**Comment by AdrianGarside** [comment]
I was wondering if you wanted to push your in progress work to main so I can update my PR and get that pushed as well?

**Comment by AdrianGarside** [comment]
It's a great check point. I need to adjust your changes a little to match my threading change so it's not a clean merge. Nothing hard but easier for me since I know what I changed :). I can take a look at the async device initialization if you wanted to work on the buttons.

**Comment by AdrianGarside** [comment]
Yep. Yeah, a dev branch that we can coordinate in and then choose when to merge that to main will make sense now we're at a somewhat functional point.

**Comment by AdrianGarside** [comment]
I was able to just merge my branch where I'd pre-merged with your changes into my PR topic branch. I'd already made that fix so that it's in my PR now. I'm going to double check the merging and then call it a night (12.30am here).

**Comment by AdrianGarside** [comment]
Ok, all looks good to go from my end.

**Comment by AdrianGarside** [comment]
I think this is not going to be a simple as I’d hoped. My plan was to just let the search for the printer repeated indefinitely but HA itself times out the initialization of the integration if it doesn’t respond in time and tries to reinitialize it, which in my naive leave the listener thread trying indefinitely resulted in a thread for each reinitialization attempt, possibly growing indefinitely if HA doesn’t have a cap on reinitialization attempts.

So I think we need to:
1. Record the printer type in the config.
2. Use this for decisions about what’s actually present rather than relying on the live response
3. Add an available state to all sensors, buttons etc. that returns false if we haven’t been able to connect to the printer. IIRC this will mark them as unavailable without them being completely gone.
4. Keep polling to try and connect to the printer or maybe find a way to get HA to trigger us off the SSDP notification the printer broadcasts every minute-ish. Us doing the polling will detect the printer much more quickly on average unless that SSDP notification fires right as the printer can be connected to, which seems likely. I’ll check.

This will keep the sensors in HA rather than it saying they don’t exist any more and you should maybe delete them. And then when the printer is finally turned on we can mark them all as available and provide starting values and I think they should all then just light up.

**Comment by AdrianGarside** [comment]
So after stepping away from this for a bit and then taking a break after the new firmware drops and forced lan mode only for mqtt, it seems this is now suddenly working. Previously I was seeing that if the firmware versions arrived after initial setup was complete the values didn't update in HA. But now they do.

I'm also seeing timeouts connecting to the mqtt server on the P1P where I didn't previously but the reconnection logic works and it successfully reconnects and gets back into a correct state. Will test an HA restart while the device is offline next.

**Comment by AdrianGarside** [comment]
And that scenario now works too. ~90s after the printer is power one, the integration can successfully connect to the MQTT and initialize all current state including the firmware versions.

**Comment by AdrianGarside** [comment]
Current stage apparently initializes to 0 after a power off:
![image](https://user-images.githubusercontent.com/26147576/224891634-136c50ce-a8be-4d35-9c6c-bb45eefca63d.png)

**Comment by AdrianGarside** [comment]
Fixed with #40
